Operator fails to patch ingress after update to 25.0.0

### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

operator

### Describe the bug

After upgrading the operator to version `25.0.0` the following error message appears in the log:

```
Caused by: io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: PATCH at: https://...:443/apis/networking.k8s.io/v1/namespaces/keycloak/ingresses/keycloak-ingress?fieldManager=keycloakcontroller&force=true. Message: Ingress.extensions "keycloak-ingress" is invalid: spec.rules[0].host: Invalid value: "https://keycloak.apps.okd.intern": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*'). Received status: Status(apiVersion=v1, code=422, details=StatusDetails(causes=[StatusCause(field=spec.rules[0].host, message=Invalid value: "https://keycloak.apps.okd.intern": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.co...
at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.requestFailure(OperationSupport.java:660)
at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.requestFailure(OperationSupport.java:640)
at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.assertResponseCode(OperationSupport.java:589)
at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.lambda$handleResponse$0(OperationSupport.java:5
```

I assume this has to do with required change in the hostname stanza:

```
  hostname:
    # used to be 
    # admin: keycloak.apps.okd.intern
    admin: https://keycloak.apps.okd.intern
    hostname: https://keycloak.apps.okd.intern
    # hostname: eycloak.apps.okd.intern
    strict: false
  ingress:
    enabled: true
    className: openshift-default
```

### Version

25.0.0

### Regression

- [X] The issue is a regression

### Expected behavior

The operator creates a valid ingress resource and has no errors during reconciliation.

### Actual behavior

The operator logs errors. The existing ingress continues to work.

### How to Reproduce?

Upgrade with `ingress.enabled` from version 24.0.X to 25.0.0.

### Anything else?

_No response_

--- a/operator/src/main/java/org/keycloak/operator/controllers/KeycloakIngressDependentResource.java
+++ b/operator/src/main/java/org/keycloak/operator/controllers/KeycloakIngressDependentResource.java
@@ -24,11 +24,14 @@ import io.javaoperatorsdk.operator.processing.dependent.kubernetes.CRUDKubernete
 import io.javaoperatorsdk.operator.processing.dependent.kubernetes.KubernetesDependent;
 import io.javaoperatorsdk.operator.processing.dependent.workflow.Condition;
 
+import io.quarkus.logging.Log;
 import org.keycloak.operator.Constants;
 import org.keycloak.operator.Utils;
 import org.keycloak.operator.crds.v2alpha1.deployment.Keycloak;
 import org.keycloak.operator.crds.v2alpha1.deployment.spec.IngressSpec;
 
+import java.net.MalformedURLException;
+import java.net.URL;
 import java.util.HashMap;
 import java.util.Optional;
 
@@ -107,7 +110,17 @@ public class KeycloakIngressDependentResource extends CRUDKubernetesDependentRes
 
         final var hostnameSpec = keycloak.getSpec().getHostnameSpec();
         if (hostnameSpec != null && hostnameSpec.getHostname() != null) {
-            ingress.getSpec().getRules().get(0).setHost(hostnameSpec.getHostname());
+            String hostname = hostnameSpec.getHostname();
+
+            try {
+                hostname = new URL(hostname).getHost();
+                Log.debug("Hostname is a URL, extracting host: " + hostname);
+            }
+            catch (MalformedURLException e) {
+                Log.debug("Hostname is not a URL, using as is: " + hostname);
+            }
+
+            ingress.getSpec().getRules().get(0).setHost(hostname);
         }
 
         return ingress;

--- a/operator/src/test/java/org/keycloak/operator/testsuite/unit/IngressLogicTest.java
+++ b/operator/src/test/java/org/keycloak/operator/testsuite/unit/IngressLogicTest.java
@@ -41,7 +41,7 @@ public class IngressLogicTest {
 
     static class MockKeycloakIngress {
 
-        private static Keycloak getKeycloak(boolean tlsConfigured, IngressSpec ingressSpec) {
+        private static Keycloak getKeycloak(boolean tlsConfigured, IngressSpec ingressSpec, String hostname) {
             var kc = K8sUtils.getDefaultKeycloakDeployment();
             kc.getMetadata().setUid("this-is-a-fake-uid");
             if (ingressSpec != null) {
@@ -49,6 +49,9 @@ public class IngressLogicTest {
             }
             if (!tlsConfigured) {
                 kc.getSpec().getHttpSpec().setTlsSecret(null);
+            }
+            if (hostname != null) {
+                kc.getSpec().getHostnameSpec().setHostname(hostname);
             }
             return kc;
         }
@@ -58,10 +61,18 @@ public class IngressLogicTest {
         }
 
         public static MockKeycloakIngress build(Boolean defaultIngressEnabled, boolean ingressExists, boolean ingressSpecDefined, boolean tlsConfigured) {
-            return build(defaultIngressEnabled, ingressExists, ingressSpecDefined, tlsConfigured, null);
+            return build(defaultIngressEnabled, ingressExists, ingressSpecDefined, tlsConfigured, null, null);
+        }
+
+        public static MockKeycloakIngress build(String hostname) {
+            return build(true, false, true, true, null, hostname);
         }
 
         public static MockKeycloakIngress build(Boolean defaultIngressEnabled, boolean ingressExists, boolean ingressSpecDefined, boolean tlsConfigured, Map<String, String> annotations) {
+            return build(defaultIngressEnabled, ingressExists, ingressSpecDefined, tlsConfigured, annotations, null);
+        }
+
+        public static MockKeycloakIngress build(Boolean defaultIngressEnabled, boolean ingressExists, boolean ingressSpecDefined, boolean tlsConfigured, Map<String, String> annotations, String hostname) {
             IngressSpec ingressSpec = null;
             if (ingressSpecDefined) {
                 ingressSpec = new IngressSpec();
@@ -72,7 +83,7 @@ public class IngressLogicTest {
                     ingressSpec.setAnnotations(annotations);
                 }
             }
-            MockKeycloakIngress mock = new MockKeycloakIngress(tlsConfigured, ingressSpec);
+            MockKeycloakIngress mock = new MockKeycloakIngress(tlsConfigured, ingressSpec, hostname);
             mock.ingressExists = ingressExists;
             return mock;
         }
@@ -82,8 +93,12 @@ public class IngressLogicTest {
         private boolean deleted = false;
         private Keycloak keycloak;
 
+        public MockKeycloakIngress(boolean tlsConfigured, IngressSpec ingressSpec, String hostname) {
+            this.keycloak = getKeycloak(tlsConfigured, ingressSpec, hostname);
+        }
+
         public MockKeycloakIngress(boolean tlsConfigured, IngressSpec ingressSpec) {
-            this.keycloak = getKeycloak(tlsConfigured, ingressSpec);
+            this(tlsConfigured, ingressSpec, null);
         }
 
         public boolean reconciled() {
@@ -101,7 +116,7 @@ public class IngressLogicTest {
                 }
                 return Optional.empty();
             }
-            
+
             return Optional.of(keycloakIngressDependentResource.desired(keycloak, null));
         }
     }
@@ -216,5 +231,13 @@ public class IngressLogicTest {
         Optional<HasMetadata> reconciled = kc.getReconciledResource();
         Ingress ingress = reconciled.map(Ingress.class::cast).orElseThrow();
         assertEquals("my-class", ingress.getSpec().getIngressClassName());
+    }
+
+    @Test
+    public void testHostnameSanitizing() {
+        var kc = MockKeycloakIngress.build("https://my-other.example.com:443/my-path");
+        Optional<HasMetadata> reconciled = kc.getReconciledResource();
+        Ingress ingress = reconciled.map(Ingress.class::cast).orElseThrow();
+        assertEquals("my-other.example.com", ingress.getSpec().getRules().get(0).getHost());
     }
 }

